src/front-end/typescript/lib/pages/feedback.tsx

```diff
@@ -0,0 +1,158 @@
+import { makePageMetadata, makeStartLoading, makeStopLoading, UpdateState } from 'front-end/lib';
+import { Route, SharedState } from 'front-end/lib/app/types';
+import { ComponentView, emptyPageAlerts, GlobalComponentMsg, newRoute, PageComponent, PageInit, Update } from 'front-end/lib/framework';
+import * as api from 'front-end/lib/http/api';
+import { updateField, validateField } from 'front-end/lib/views/form-field';
+import Icon from 'front-end/lib/views/icon';
+import * as LongText from 'front-end/lib/views/input/long-text';
+import FixedBar from 'front-end/lib/views/layout/fixed-bar';
+import Link from 'front-end/lib/views/link';
+import LoadingButton from 'front-end/lib/views/loading-button';
+import { default as React } from 'react';
+import { Col, Row } from 'reactstrap';
+import { ADT, Rating } from 'shared/lib/types';
+import { validateGenericString, validateRating } from 'shared/lib/validators';
+import { validateFeedbackText } from 'shared/lib/validators/feedback';
+
+export interface State {
+  loading: number;
+  rating?: Rating;
+  feedbackText: LongText.State
+}
+
+type InnerMsg
+  = ADT<'onChangeRating', Rating>
+  | ADT<'onChangeFeedbackText', string>
+  | ADT<'validateFeedback'>
+  | ADT<'submit'>;
+
+export type Msg = GlobalComponentMsg<InnerMsg, Route>;
+
+export type RouteParams = null;
+
+const init: PageInit<RouteParams, SharedState, State, Msg> = async () => {
+  return {
+    loading: 0,
+    rating: undefined,
+    feedbackText: LongText.init({
+      id: 'feedback-text',
+      required: false,
+      placeholder: 'Describe your experience here.',
+      label: 'Please describe your experience.'
+    })
+  };
+};
+
+const startLoading: UpdateState<State> = makeStartLoading('loading');
+const stopLoading: UpdateState<State> = makeStopLoading('loading');
+
+const update: Update<State, Msg> = ({ state, msg }) => {
+  switch (msg.tag) {
+    case 'onChangeRating':
+      return [state.set('rating', msg.value)];
+    case 'onChangeFeedbackText':
+      return [updateField(state, 'feedbackText', msg.value)];
+    case 'validateFeedback':
+      return [validateField(state, 'feedbackText', validateFeedbackText)]
+    case 'submit':
+      state = startLoading(state);
+      return [
+        state,
+        async (state, dispatch) => {
+          const validatedRating = validateRating(state.rating);
+          if (validatedRating.tag === 'invalid') {
+            return state;
+          }
+          await api.createFeedback({
+            rating: validatedRating.value,
+            text: state.feedbackText.value
+          });
```

Comment:
Need to handle the failure case of this HTTP request. What happens if this request fails, or if there is a server-side validation error?